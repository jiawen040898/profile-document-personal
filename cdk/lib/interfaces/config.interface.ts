export type CommonCDKEnvironmentVariables = {
    DOCUMENT_UPLOAD_BUCKET_NAME: string;
    PULSIFI_ASSETS_BUCKET: string;
    PULSIFI_ASSETS_DOMAIN: string;
    REGION: string;
    SENTRY_DSN: string;
    DAXTRA_API_URL: string;
    DAXTRA_ACCOUNT: string;
    DAXTRA_JWT_SECRET: string;
    DAXTRA_API_TIMEOUT: string;
    SERVERLESS_STAGE: string;
    AWS_ALB_BASE_DNS: string;
};

export type CDKEnvironmentVariables = {
    FILE_SCANNER_ENABLED: string;
};

export type ProfileDocumentResumeParserFnConfig = {
    UNLEASH_API_KEY: string;
    UNLEASH_API_URL: string;
    UNLEASH_ENV: string;
    UNLEASH_PROJECT_ID: string;
};
