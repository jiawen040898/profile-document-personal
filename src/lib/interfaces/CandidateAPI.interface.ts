export interface ICandidateAddResume {
    file_name: string;
    file_path: string;
    file_type: string;
    content_path: string;
}
