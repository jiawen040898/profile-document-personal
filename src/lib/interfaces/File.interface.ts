export interface File {
    filename: string;
    mimeType?: string;
    encoding?: string;
    truncated?: boolean;
    content: Buffer;
}

export interface ConvertPdfPayload {
    sourceBucket: string;
    sourceKey: string;
}

export interface UploadingFileResponseDto {
    file_path: string;
    file_name: string;
    bucket_name: string;
    key: string;
    file_extension?: string;
}
