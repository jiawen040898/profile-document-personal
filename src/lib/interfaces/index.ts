export * from './CandidateAPI.interface';
export * from './File.interface';
export * from './PartnerResumeUpload.interface';
export * from './Resume.interface';
export * from './ResumeUpload.interface';
export * from './UploadOutputs.interface';
export * from './UploadOutputs.interface';
