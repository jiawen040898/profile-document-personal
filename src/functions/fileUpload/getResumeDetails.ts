import { OutputParsedContent, Resume, UploadOutput } from '@pulsifi/interfaces';
import { resumeMapper } from '@pulsifi/mappers';
import { parseResume } from '@pulsifi/services';

export async function getResumeDetails(
    file: Buffer,
    output: UploadOutput,
): Promise<{
    fullDetails: Resume;
    partialDetails: OutputParsedContent;
}> {
    const b64File: string = file.toString('base64');
    const fullDetails: Resume = await parseResume(b64File, output);

    const getMappedResumeDetails = resumeMapper.toResumeDetails(
        fullDetails,
        '',
    );

    const partialDetails: OutputParsedContent = {
        profile: getMappedResumeDetails.parsed_content.profile,
        careers: getMappedResumeDetails.parsed_content.careers,
        skills: getMappedResumeDetails.parsed_content.skills,
    };

    return {
        fullDetails,
        partialDetails,
    };
}
