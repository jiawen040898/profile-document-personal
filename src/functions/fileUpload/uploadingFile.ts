import { PulsifiConfig, S3Config } from '@pulsifi/configs';
import { logger } from '@pulsifi/fn';
import { File, UploadingFileResponseDto } from '@pulsifi/interfaces';
import { copyFile, uploadPublicItem } from '@pulsifi/services';
import { FileUtil } from '@pulsifi/utils';
import { NIL as NIL_UUID } from 'uuid';

import { CustomError, ErrorType } from '../../lib/CustomError';

const DocumentUploadBucketName = S3Config.bucket_name;
const PulsifiAssetsBucketName = S3Config.assets_bucket_name;
const ClientAssetBasedPath = 'client';

export async function uploadingFile(
    file: File,
    shouldCheckFileScannedResult: boolean,
    uuid: string = NIL_UUID,
    intendedFileName = 'document',
    basePath = 'candidate',
): Promise<UploadingFileResponseDto> {
    const fileExtension = FileUtil.getFileExtension(file.filename);
    const fileName = `${intendedFileName}.${fileExtension}`;
    const fileNameWithDirectory = `${basePath}/${uuid}/${fileName}`;

    try {
        await uploadPublicItem(
            DocumentUploadBucketName,
            fileNameWithDirectory,
            file.content,
            FileUtil.getMimeType(file),
            shouldCheckFileScannedResult,
        );
    } catch (error) {
        if (error instanceof CustomError) {
            throw error;
        }

        logger.error('Failed to upload file to s3.', {
            context: 'S3',
            bucket: DocumentUploadBucketName,
            key: fileNameWithDirectory,
        });

        throw CustomError.get(ErrorType.INTERNAL_SERVER_ERROR, error);
    }

    return {
        file_path: `https://${DocumentUploadBucketName}.s3-${S3Config.aws_region}.amazonaws.com/${fileNameWithDirectory}`,
        file_name: fileName,
        bucket_name: DocumentUploadBucketName,
        key: fileNameWithDirectory,
        file_extension: fileExtension,
    };
}

export async function copyFileToCompanyAssetFolder(
    uploadedFile: UploadingFileResponseDto,
    companyIdPath: number | string,
    uuid: string,
): Promise<UploadingFileResponseDto> {
    const fileExtension = uploadedFile.file_extension;
    const sourceBucket = uploadedFile.bucket_name;
    const sourceKey = uploadedFile.key;
    const fileName = `${uuid}.${fileExtension}`;
    const fileNameWithDirectory = `${ClientAssetBasedPath}/${companyIdPath}/${fileName}`;

    await copyFile(
        sourceBucket,
        sourceKey,
        PulsifiAssetsBucketName,
        fileNameWithDirectory,
    );

    return {
        file_path: `${PulsifiConfig.pulsifi_assets_domain}/${fileNameWithDirectory}`,
        file_name: fileName,
        bucket_name: PulsifiAssetsBucketName,
        key: fileNameWithDirectory,
        file_extension: fileExtension,
    };
}
