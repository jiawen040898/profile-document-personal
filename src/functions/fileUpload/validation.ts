import {
    FileExtension,
    FilePurpose,
    ResumeFileExtensions,
} from '@pulsifi/enum';
import { logger } from '@pulsifi/fn';
import { File } from '@pulsifi/interfaces';

import { CustomError, ErrorType } from '../../lib/CustomError';

export function validateFile(filePurpose: FilePurpose, file: File): void {
    const matches = /(?:\.([^.]+))?$/.exec(file.filename);
    const fileExtension = `.${matches![1]}`.toLowerCase();

    logger.info('file extension: ', {
        matches,
        file_purpose: filePurpose,
        filename: file.filename,
        fileExtension,
    });

    if (
        !file ||
        !Object.values(FilePurpose).includes(filePurpose) ||
        !file.content
    ) {
        throw CustomError.get(ErrorType.INVALID_UPLOAD_INPUT);
    }

    if (
        filePurpose === FilePurpose.RESUME &&
        !Object.values(ResumeFileExtensions).includes(fileExtension)
    ) {
        throw CustomError.get(ErrorType.INVALID_UPLOAD_FILE_EXTENSION);
    }

    if (
        filePurpose === FilePurpose.ATTACHMENT &&
        !Object.values(FileExtension).includes(fileExtension as FileExtension)
    ) {
        throw CustomError.get(ErrorType.INVALID_UPLOAD_FILE_EXTENSION);
    }
}
