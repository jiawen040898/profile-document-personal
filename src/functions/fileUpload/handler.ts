import { FilePurpose } from '@pulsifi/enum';
import { generatorUtil } from '@pulsifi/fn';
import { File, OutputResume, UploadOutput } from '@pulsifi/interfaces';

import { httpMiddleware } from '../../lib/commonMiddleware';
import { getResumeDetails } from './getResumeDetails';
import { uploadingFile } from './uploadingFile';
import { validateFile } from './validation';

// endpoint input: {file: file, file_purpose: attachment|resume}

// header Content-Type have to set to multipart/form-data
const uploadHandler = async (event: SafeAny /*, context*/) => {
    // eslint-disable-next-line @typescript-eslint/naming-convention
    const { file_purpose, file, company_id, company_name, job_id, job_title } =
        event.body;

    validateFile(file_purpose, file);

    const uuid = generatorUtil.uuid();

    const savedItem = await uploadingFile(
        file,
        file_purpose === FilePurpose.RESUME ||
            file_purpose === FilePurpose.ATTACHMENT,
        uuid,
        file_purpose,
    );

    const output: UploadOutput = {
        file_path: savedItem.file_path,
        file_name: savedItem.file_name,
        meta: {
            company_id,
            company_name,
            job_id,
            job_title,
        },
    };

    if (file_purpose == FilePurpose.RESUME) {
        const { fullDetails, partialDetails } = await getResumeDetails(
            file.content,
            output,
        );
        const fullDetailsJson = JSON.stringify(fullDetails);
        const fullDetailsBuffer = Buffer.from(fullDetailsJson, 'utf-8');
        const fullDetailsJsonFile: File = {
            content: fullDetailsBuffer,
            filename: 'daxtra.json',
            mimeType: 'application/json',
        };
        const savedJson = await uploadingFile(
            fullDetailsJsonFile,
            false,
            uuid,
            'daxtra',
        );

        const outputResume: OutputResume = {
            content_path: savedJson.file_path,
            parsed_content: partialDetails,
        };

        output.resume = outputResume;
    }

    return {
        statusCode: 200,
        body: JSON.stringify({ data: output }, null, 2),
    };
};

export const handler = httpMiddleware(uploadHandler);
