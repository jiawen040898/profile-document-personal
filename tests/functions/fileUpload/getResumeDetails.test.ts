import { UploadOutput } from '@pulsifi/interfaces';
import { readFileSync } from 'fs';
import path from 'path';

import { getResumeDetails } from '../../../src/functions/fileUpload/getResumeDetails';

jest.mock('../../../src/lib/services/daxtra.service');

const content: Buffer = readFileSync(
    path.join(__dirname, '../../fixtures/sample-pdf.pdf'),
);

describe('getResumeDetails', () => {
    it('should able to run without error', async () => {
        const resumeOutput = await getResumeDetails(
            content,
            {} as UploadOutput,
        );

        expect(resumeOutput.fullDetails).toBeTruthy();
        expect(resumeOutput.partialDetails).toMatchSnapshot();
    });
});
