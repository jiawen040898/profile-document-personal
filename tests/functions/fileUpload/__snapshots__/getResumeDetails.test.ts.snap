// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getResumeDetails should able to run without error 1`] = `
{
  "careers": [
    {
      "country_code": null,
      "description": "*    Develop new features based on requirement (full-stack)
    *    Fix urgent bugs in the system.
    *    Breaking down big tasks and assigning them to new recruits.
    *    Participated in maintenance of the system that handles 10,000,000+ API calls per day.
    *    Wrote a report generator that analyzes and summarizes 25,000,000 mongo documents within a minute. Some parts of its functions are also being reused for other purposes later by colleagues, which massively shortens their development process.
    *    Creating a project to scrape data from third party websites through Python and Scrapy, automate them with UI for setting and data presentation.
    *    Designing and adjusting databases based on requirement and performance issues.
    *    Use selenium to mimic user behavior for testing.
    *    Using Pyautogui to automate tasks for the company.
    *    Keep exploring new ways to achieve the company needs.",
      "end_date": null,
      "is_current": true,
      "months_of_work": 45,
      "organization": "Johor",
      "role": "Javascript Developer",
      "start_date": "2017-05",
    },
    {
      "country_code": null,
      "description": "*    Add Js and PHP on top of Wordpress to customize pages.
    *    Add modules on top of the current Laravel system in the company to fulfill company needs.
    *    Adjust the way Laravel handles upload so that it will add watermark and compress image size on PDF, GIF, PNG, BMP and JPG.

Freelancing, Remote - JS & PHP Developer",
      "end_date": "2017-04",
      "is_current": false,
      "months_of_work": 2,
      "organization": "K.L.",
      "role": "Junior Software Developer",
      "start_date": "2017-03",
    },
    {
      "country_code": null,
      "description": "*    Wrote a set of Javascript image handlers that do cropping, resizing and rotation of image.
    *    Built a PHP backend API for a mobile app for gym to allow all their customers to check-in through QR codes for virtual currencies which can exchange for physical rewards.
    *    Single handedly built a PHP backend API for a Jewellery Reseller to track their stocks flow, and transaction records. Actions of all users are tracked for the security purposes.",
      "end_date": "2016-10",
      "is_current": false,
      "months_of_work": 16,
      "organization": null,
      "role": null,
      "start_date": "2015-07",
    },
  ],
  "profile": {
    "email": "huatzhi@gmail.com",
    "first_name": "Lor",
    "last_name": "Zhi",
    "nationality": null,
    "phone_code": "60",
    "phone_number": "1110898231",
    "professional_summary": null,
  },
  "skills": [
    {
      "name": "Chinese",
      "proficiency": null,
      "source": "daxtra",
    },
    {
      "name": "English",
      "proficiency": null,
      "source": "daxtra",
    },
  ],
}
`;
